### PR TITLE
Model debugger upgrades

### PR DESCRIPTION
**Pull request by molbap**
# What does this PR do?

A continuation of #36798 , now:

- The debugger will only output the first and last layer of a sequence of layers.
- mean/stds are added as well, and a `..._SUMMARY.json` file will contain only statistics, not full tensors.
- General printing improvements.

### CODE REVIEW

File: src/transformers/model_debugging_utils.py [hunk 1]
```diff
@@ -87,20 +88,51 @@ def _serialize_io(value):
 
     if hasattr(value, "_local_tensor"):
         # DTensor-like handling, just use local tensor attribute
-        return {
+        torch.set_printoptions(sci_mode=True)
```
Comment (ArthurZucker) [inline src/transformers/model_debugging_utils.py:91]: we could ahve max line width increased as well1

**Review by ArthurZucker — APPROVED** [review]
Nice!

**Comment by ArthurZucker** [comment]
Can we make sure hooks are removed after we exit the context manager?

File: src/transformers/model_debugging_utils.py (hunk 1 above)
Comment (molbap) [reply src/transformers/model_debugging_utils.py:91]: it's done in the `repr_to_list` method, will unify this ;)

**Comment by molbap** [comment]
cc @eustlb @zucchini-nlp @qubvel @yonigozlan , my fellow model adders, could be helpful! Check out the doc (and @Cyrilvallez but I think you've seen/used it already maybe)

**Comment by Cyrilvallez** [comment]
Hey @molbap, I just had a random thought while reviewing a model and was thinking about your super nice util, so dropping it here right now as I'm seeing the tag (this is NOT intended as a follow-up to your util hahaha, just to see if it could be helpful for everyone/others have ideas about that):

Another super cool helper IMO would be some kind of library scanner to find close models for modular. E.g., I want to find a MLP with only 2 Linear layers and Dropout -> helper find the different related models
As far as I know we cannot easily do it with usual IDE search tools as most of the time we don't know (and don't really care, as we have a weight converter most of the time anyway) the actual names of those layers

**Comment by qubvel** [comment]
Following @CyrilVallez's idea, it might indeed be very helpful to run such a tool even across the library to find identical modules that may have differently named attributes but are identical actually. Unfortunately, we cannot rename attributes because it would break the weights loading (or we would need a hook for renaming), but at least we can choose one as a standard and leave comments for the rest, indicating that e.g. LlamaMLP is identical to this one and is preferred to be used for modular purposes

File: tests/utils/test_model_debugging_utils.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import tempfile
+import unittest
+from pathlib import Path
+
+from transformers import is_torch_available, is_vision_available
+from transformers.model_debugging_utils import model_addition_debugger_context
+
+
+if is_vision_available():
+    pass
```
Comment (qubvel) [inline tests/utils/test_model_debugging_utils.py:28]: can be removed

File: tests/utils/test_model_debugging_utils.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+if is_torch_available():
+    import torch
+    from torch import nn
+
+
+class ToyModel(nn.Module):
```
Comment (qubvel) [inline tests/utils/test_model_debugging_utils.py:35]: Should be also under the `if is_torch_available()`, otherwise we don't need a guard actually

File: docs/source/en/internal/model_debugging_utils.md [hunk 4]
```diff
@@ -66,6 +67,141 @@ with model_addition_debugger_context(model, "optional_path_to_your_output_file.j
 ```
 
 
-[[autodoc]] model_addition_debugger
+### Reading results
```
Comment (qubvel) [inline docs/source/en/internal/model_debugging_utils.md:70]: Comment for the above lines actually, but can't comment there.

```
# call forward method (not .generate!)
with model_addition_debugger_context(model, "optional_path_to_your_output_file.json"):
    output = model.forward(**inputs)
```

Should we call `model.forward`? Why not `model(**intpus)`? Are we avoiding the top-level hooks for some reason?

**Review by qubvel — COMMENTED** [review]
Nice, thanks for updating!

> The debugger will only output the first and last layer of a sequence of layers.

Is it a configurable option? I remember for mllama we messed up self/cross attention layers order, so the diff appears after layer 4.

File: docs/source/en/internal/model_debugging_utils.md (hunk 4 above)
Comment (molbap) [reply docs/source/en/internal/model_debugging_utils.md:70]: no, top-level works as well - it's just to be explicit and oppose `model.forward` to `model.generate`! Can explain (as the latter would create a several 100MB json file...)

File: tests/utils/test_model_debugging_utils.py (hunk 2 above)
Comment (molbap) [reply tests/utils/test_model_debugging_utils.py:28]: indeed, artifact of `make fixup`

**Comment by molbap** [comment]
Added:
- configurable `do_prune_layers` and associated test @qubvel (+ torch guarding the whole thing, why not)
- more documentation
- docstrings because I'm feeling nice today

image

merging!
